src/utils/pceimg/pce-img.1
.TH PCE-IMG 1 "2012-01-30" "HH" "pce"
\
.SH NAME
pce-img \- the PCE disk image utility

.SH SYNOPSIS
.BI "pce-img create" " [options] [output-file]"

.BI "pce-img convert" " [options] [input-file] [options] [output-file]"

.SH DESCRIPTION
\fBpce-img\fR(1) is used to create and convert disk image files.

.SH OPTIONS
.TP
.BI "-c, --cylinders " c
Set the number of cylinders.
\
.TP
.BI "-C, --min-cluster-size " size
Set the minimum cluster size for QED images. The default is 0.
\
.TP
.BI "-f, --offset " offset
Set the starting offset of the disk image for formats that
support it.
\
.TP
.BI "-g, --geometry " "c h s"
Set the disk geometry. This is the same as using the \fB-c\fR,
\fB-h\fR and \fB-s\fR options.
\
.TP
.BI "-h, --heads " h
Set the number of heads.
\
.TP
.BI "-i, --input " filename
Set the input file name.
\
.TP
.BI "-m, --megabytes " size
Set the disk size in MiB.
\
.TP
.BI "-n, --size " blocks
Set the disk size in 512 byte blocks.
\
.TP
.BI "-o, --output " filename
Set the output file name.
\
.TP
.B "-q, --quiet"
Be quiet.
\
.TP
.BI "-s, --sectors " s
Set the number of sectors per track.
\
.TP
.BI "-w, --cow " filename
Set the COW file name.
\
.TP
.B --help
Print usage information.
\
.TP
.B --version
Print version information.

.SH EXAMPLES
Create a 1 GiB PIMG image:
.IP
$ pce-img create -m 1024 -o image.pimg
.PP
Create a 10 MB PIMG image
.IP
$ pce-img create -g 306 4 17 -o image.pimg
.PP
Convert a PIMG image to a raw image
.IP
$ pce-img convert source.pimg dest.img

.SH SEE ALSO
.BR pce-ibmpc "(1),"
.BR pce-macplus "(1),"
.BR pfdc "(1)"

.SH AUTHOR
Hampa Hug <hampa@hampa.ch>


NEWS
0.2.1	2011-09-06

	* ibmpc: Better support the PC/XT 5160.
	* ibmpc: Add a new monitor command "log int" to log interrupt calls.
	* ibmpc: Add FDC emulation.
	* ibmpc: The CPU speed can now be set to 0, meaning "all out".
	* ibmpc: Experimental sound support (speaker only).
	* macplus: Add Macintosh SE emulation.
	* macplus: Add Macintosh Classic emulation.
	* macplus: Experimental sound support.
	* macplus: Reimplement the sony driver.
	* 8086: Fix the int 0 return address.
	* 8086: Fix overflow detection for the IDIV and DIV instructions.
	* config: A complete rewrite of the config file parser.
	* block: Add a new block driver which handles PCE (.pfdc),
	  Teledisk (.td0) and ImageDisk (.imd) floppy disk image files.
	* block: Add "pfdc", a program to manipulate floppy disk image files.
	* block: Add a new block driver for QED image files.
	* block: Add support for Apple Disk Copy 4.2 image files.
	* cga: Support the CGA start address in graphics modes.
	* vga: Support the vertical retrace interrupt.
	* vga: Support the IBM VGA BIOS.
	* sdl: Fix SDL mouse grabbing.
	* sdl: Fix entering and leaving full screen mode.

0.2.0	2009-11-30

	* Add a Macintosh Plus emulator.
	* Add emulation of the IBM PC cassette interface.
	* Serial and parallel ports are now connected to generic character
	  drivers.
	* Add VGA emulation. The emulated VGA card is a plain IBM VGA without
	  extras.
	* Add Wyse-700 emulation.
	* Support HGC character height in graphics mode. This supports
	  odd HGC graphic modes like 640*300 that are sometimes used
	  by games.
	* The MDA and HGC screen colors can now be configured.
	* Support blinking cursor and characters. The blink rate can
	  be set in the config file.
	* Fix a bug in the 80186 emulation that lead some programs
	  to detect the CPU as an 80286.
	* The PC BIOS is now patched at runtime which means that a completely
	  unmodified PC or XT BIOS should work.
	* Remove the half disassembled BIOS source. A binary BIOS ROM image
	  is now required.
	* All config file options that specify a number now accept the
	  suffixes "K", "M" and "G" to multiply the number by 1024, 1048576
	  and 1073741824, respectively.
	* The monitor interface now works under Windows.

0.1.8	2008-11-07	r863

	* A complete rewrite of the terminal code.
	* Removed the VT100 terminal.
	* Improved video device emulation.
	* Emulated time is now synchronized with real time.
	* Lots of small changes and bugfixes.

0.1.7	2006-07-24	r663

	* Implemented Motorola S-Record loading support.
	* Added 'pceimg', an application to create and convert disk images.
	* Added an 68000 core emulator.
	* Bugfixes

0.1.6	2005-02-28	r540

	* Added an ARM core emulator and a minimal system simulator
	* Added a sparc32 core emulator and a minimal system simulator
	* Cleaned up the block device code.
	* Added a new partitioned block device type.
	* Added copy on write support to all block devices.
	* Added a new pce specific block device type (pimg).
	* Added a new DOS utility pcemsg.com to send messages to the
	  emulation core.
	* Bugfixes

0.1.5	2004-08-03	r439

	* Added a 6502 emulator
	* Added minimal architectures for the PowerPC 405 and 6502
	  emulators.
	* Added a VGA emulator. Partially functional but not yet complete.
	* Fixed the EGA palette
	* Improved the mouse emulation. More mouse drivers should
	  work now.
	* Cleaned up terminal handling
	* Added support for intel hex files.

0.1.4	2004-02-22	r387

	* Added HMA support
	* Added EMS 3.2 support
	* Switched from CVS to SVN
	* Reverted to old file-only serial port
	* Added a DMAC 8237 emulator
	* Added a PowerPC 405 emulator

0.1.3	2003-10-06	r287

	* EGA emulation is now much more complete
	* CGA graphic modes are now supported on EGA

0.1.2	2003-09-23	r257

	* Added an SDL terminal
	* Added some crude EGA support

0.1.1	2003-09-04	r193

	* Added a UART 8250 emulator
	* Added serial port support
	* Added mouse support
	* Added XMS 3.0 support, including UMB's and a DOS HIMEM driver
	* Bugfixes

0.1.0	2003-08-20	r142

	* Added parallel port support
	* Added hercules graphics card support
	* Added graphic support to the X terminal
	* CGA and HGC graphic modes are supported

0.0.6	2003-04-27	r123

	* Added a PIT 8253 emulator
	* Massive source tree restructuring
	* The trap flag is now handled almost correctly
	* Completely reworked terminal support
	* Added X11 terminal support
	* CPU identification should now work

0.0.5	2003-04-23	r82

	* Completed instruction timings
	* Added a PIC 8259A emulator
	* Added config file parsing

0.0.4	2003-04-19	r51

	* Added ESC opcodes
	* Added lots of instruction timings
	* Fixed a serious bug in the LDS/LES instructions
	* The overflow flag is now set correctly
	* Added a PPI 8255 emulator
	* Added a CGA emulator (text mode only)
	* Added support for DOSEMU hdimages
	* The main application now uses command line options
	* Lots of small bugfixes

0.0.3	2003-04-15	r4

	* Lots of changes. Basically a new project.
	* First cvs version


0.0.2	2002-05-20

0.0.1	2001-05-02


src/utils/pceimg/pce-img.c
/*****************************************************************************
 * pce                                                                       *
 *****************************************************************************/

/*****************************************************************************
 * File name:   src/utils/pceimg/pce-img.c                                   *
 * Created:     2005-11-29 by Hampa Hug <hampa@hampa.ch>                     *
 * Copyright:   (C) 2005-2012 Hampa Hug <hampa@hampa.ch>                     *
 *****************************************************************************/

/*****************************************************************************
 * This program is free software. You can redistribute it and / or modify it *
 * under the terms of the GNU General Public License version 2 as  published *
 * by the Free Software Foundation.                                          *
 *                                                                           *
 * This program is distributed in the hope  that  it  will  be  useful,  but *
 * WITHOUT  ANY   WARRANTY,   without   even   the   implied   warranty   of *
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU  General *
 * Public License for more details.                                          *
 *****************************************************************************/


#include <config.h>

#include "pce-img.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include <drivers/block/block.h>
#include <drivers/block/blkcow.h>
#include <drivers/block/blkfdc.h>
#include <drivers/block/blkraw.h>
#include <drivers/block/blkpce.h>
#include <drivers/block/blkqed.h>
#include <drivers/block/blkdosem.h>

#include <lib/getopt.h>


#define DSK_NONE   0
#define DSK_RAW    1
#define DSK_PCE    2
#define DSK_DOSEMU 3
#define DSK_PFDC   4
#define DSK_QED    5


static const char    *argv0 = NULL;

static int           par_quiet = 0;

static unsigned long par_min_cluster_size = 0;


static pce_option_t opts_main[] = {
	{ '?', 0, "help", NULL, "Print usage information" },
	{ 'V', 0, "version", NULL, "Print version information" },
	{  -1, 0, NULL, NULL, NULL }
};

static pce_option_t opts_create[] = {
	{ 'c', 1, "cylinders", "int", "Set the number of cylinders [0]" },
	{ 'C', 1, "min-cluster-size", "int", "Set the minimum cluster size for QED [0]" },
	{ 'f', 1, "offset", "int", "Set the data offset [0]" },
	{ 'g', 3, "geometry", "3*int", "Set the disk geometry (c h s)" },
	{ 'h', 1, "heads", "int", "Set the number of heads [0]" },
	{ 'm', 1, "megabytes", "int", "Set the disk size in megabytes [0]" },
	{ 'n', 1, "size", "int", "Set the disk size in 512 byte blocks [0]" },
	{ 'o', 1, "output", "string", "Set the output file name [stdout]" },
	{ 'q', 0, "quiet", NULL, "Be quiet [no]" },
	{ 's', 1, "sectors", "int", "Set the number of sectors per track [0]" },
	{  -1, 0, NULL, NULL, NULL }
};

static pce_option_t opts_convert[] = {
	{ 'c', 1, "cylinders", "int", "Set the number of cylinders [0]" },
	{ 'C', 1, "min-cluster-size", "int", "Set the minimum cluster size for QED [0]" },
	{ 'f', 1, "offset", "int", "Set the data offset [0]" },
	{ 'g', 3, "geometry", "3*int", "Set the disk geometry (c h s)" },
	{ 'h', 1, "heads", "int", "Set the number of heads [0]" },
	{ 'i', 1, "input", "string", "Set the input file name [stdin]" },
	{ 'm', 1, "megabytes", "int", "Set the disk size in megabytes [0]" },
	{ 'n', 1, "size", "int", "Set the disk size in 512 byte blocks [0]" },
	{ 'o', 1, "output", "string", "Set the output file name [stdout]" },
	{ 'q', 0, "quiet", NULL, "Be quiet [no]" },
	{ 's', 1, "sectors", "int", "Set the number of sectors per track [0]" },
	{ 'w', 1, "cow", "string", "Set the COW file name [none]" },
	{  -1, 0, NULL, NULL, NULL }
};


static
void print_help (void)
{
	pce_getopt_help (
		"pce-img: PCE disk image utility",
		"usage: pce-img [options] command [options]",
		opts_main
	);

	fputs ("\n", stdout);

	pce_getopt_help (
		NULL,
		"usage: pce-img create [options] [output]",
		opts_create
	);

	fputs ("\n", stdout);

	pce_getopt_help (
		NULL,
		"usage: pce-img convert [options] [input [output]]",
		opts_convert
	);

	fputs (
		"\nfile names: <format>:<name>\n"
		"formats:    raw, pce, dosemu, pfdc, qed\n",
		stdout
	);

	fflush (stdout);
}

static
void print_version (void)
{
	fputs (
		"pce-img version " PCE_VERSION_STR
		"\n\n"
		"Copyright (C) 2005-2012 Hampa Hug <hampa@hampa.ch>\n",
		stdout
	);

	fflush (stdout);
}


/*
 * Guess the image type based on the file name extension.
 */
static
unsigned pce_get_disk_type_ext (const char *str)
{
	const char *ext;

	ext = NULL;

	while (*str != 0) {
		if (*str == '.') {
			ext = str + 1;
		}

		str += 1;
	}

	if (ext == NULL) {
		return (DSK_NONE);
	}

	if (strcasecmp (ext, "img") == 0) {
		return (DSK_RAW);
	}
	else if (strcasecmp (ext, "pimg") == 0) {
		return (DSK_PCE);
	}
	else if (strcasecmp (ext, "pfdc") == 0) {
		return (DSK_PFDC);
	}
	else if (strcasecmp (ext, "qed") == 0) {
		return (DSK_QED);
	}
	else if (strcasecmp (ext, "raw") == 0) {
		return (DSK_RAW);
	}

	return (DSK_NONE);
}

static
unsigned pce_get_disk_type (const char *str)
{
	unsigned i;
	char     buf[256];

	i = 0;
	while (str[i] != ':') {
		if (str[i] == 0) {
			return (pce_get_disk_type_ext (str));
		}

		if (i >= 256) {
			return (DSK_NONE);
		}

		buf[i] = str[i];

		i += 1;
	}

	buf[i] = 0;

	if ((strcmp (buf, "raw") == 0) || (strcmp (buf, "img") == 0)) {
		return (DSK_RAW);
	}

	if ((strcmp (buf, "pce") == 0) || (strcmp (buf, "pimg") == 0)) {
		return (DSK_PCE);
	}

	if (strcmp (buf, "qed") == 0) {
		return (DSK_QED);
	}

	if (strcmp (buf, "dosemu") == 0) {
		return (DSK_DOSEMU);
	}

	if (strcmp (buf, "pfdc") == 0) {
		return (DSK_PFDC);
	}

	return (DSK_NONE);
}

static
const char *pce_get_disk_name (const char *str)
{
	unsigned i;

	i = 0;
	while (str[i] != ':') {
		if (str[i] == 0) {
			return (str);
		}

		i += 1;
	}

	return (str + i + 1);
}

static
int pce_block_is_null (const void *buf, unsigned cnt)
{
	unsigned            i;
	const unsigned char *tmp;

	tmp = buf;

	for (i = 0; i < cnt; i++) {
		if (tmp[i] != 0) {
			return (0);
		}
	}

	return (1);
}

static
int dsk_create (const char *str, uint32_t n, uint32_t c, uint32_t h, uint32_t s,
	uint64_t ofs)
{
	unsigned   type;
	const char *name;

	type = pce_get_disk_type (str);
	name = pce_get_disk_name (str);

	switch (type) {
	case DSK_RAW:
		return (dsk_img_create (name, n, ofs));

	case DSK_PCE:
		return (dsk_pce_create (name, n, c, h, s, ofs & 0xffffffff));

	case DSK_QED:
		return (dsk_qed_create (name, n, par_min_cluster_size));

	case DSK_DOSEMU:
		return (dsk_dosemu_create (name, c, h, s, ofs & 0xffffffff));

	case DSK_PFDC:
		return (1);
	}

	return (dsk_pce_create (name, n, c, h, s, ofs & 0xffffffff));
}

static
disk_t *dsk_open (const char *str, uint32_t n, uint32_t c, uint32_t h, uint32_t s,
	uint64_t ofs, int ro)
{
	unsigned   type;
	const char *name;

	type = pce_get_disk_type (str);
	name = pce_get_disk_name (str);

	switch (type) {
	case DSK_RAW:
		return (dsk_img_open (name, ofs, ro));

	case DSK_PCE:
		return (dsk_pce_open (name, ro));

	case DSK_QED:
		return (dsk_qed_open (name, ro));

	case DSK_DOSEMU:
		return (dsk_dosemu_open (name, ro));

	case DSK_PFDC:
		return (dsk_fdc_open_pfdc (name, ro));
	}

	return (dsk_auto_open (name, ofs, ro));
}

static
int dsk_copy (disk_t *dst, disk_t *src)
{
	uint32_t      i, n, m;
	uint32_t      prg_i, prg_n;
	unsigned      k;
	uint16_t      msk;
	unsigned char buf[8192];

	n = dsk_get_block_cnt (dst);
	m = dsk_get_block_cnt (src);

	if (m < n) {
		n = m;
	}

	prg_i = 0;
	prg_n = n;

	i = 0;
	while (n > 0) {
		m = (n < 16) ? n : 16;

		if (dsk_read_lba (src, buf, i, m)) {
			return (1);
		}

		msk = 0;
		for (k = 0; k < m; k++) {
			if (pce_block_is_null (buf + 512 * k, 512) == 0) {
				msk |= 1U << k;
			}
		}

		if (msk == 0xffff) {
			if (dsk_write_lba (dst, buf, i, m)) {
				return (1);
			}
		}
		else if (msk != 0) {
			for (k = 0; k < m; k++) {
				if (msk & (1U << k)) {
					if (dsk_write_lba (dst, buf + 512 * k, i + k, 1)) {
						return (1);
					}
				}
			}
		}

		i += m;
		n -= m;

		if (par_quiet == 0) {
			prg_i += m;
			if (prg_i >= 4096) {
				fprintf (stdout, "[%6.2f%%] block %lu of %lu\r",
					(100.0 * (i + 1)) / prg_n,
					(unsigned long) i,
					(unsigned long) prg_n
				);

				prg_i = 0;

				fflush (stdout);
			}
		}
	}

	if (par_quiet == 0) {
		fprintf (stdout, "[%6.2f%%] block %lu of %lu\n",
			100.0, (unsigned long) prg_n, (unsigned long) prg_n
		);
	}

	return (0);
}

static
int main_create (int argc, char **argv)
{
	int      r;
	char     **optarg;
	char     *par_out = NULL;
	uint32_t par_c = 0;
	uint32_t par_h = 0;
	uint32_t par_s = 0;
	uint32_t par_n = 0;
	uint64_t par_ofs = 0;

	while (1) {
		r = pce_getopt (argc, argv, &optarg, opts_convert);

		if (r == GETOPT_DONE) {
			break;
		}

		if (r < 0) {
			return (1);
		}

		switch (r) {
		case '?':
			print_help();
			return (0);

		case 'V':
			print_version();
			return (0);

		case 'c':
			par_c = strtoul (optarg[0], NULL, 0);
			break;

		case 'C':
			par_min_cluster_size = strtoul (optarg[0], NULL, 0);
			break;

		case 'f':
			par_ofs = strtoull (optarg[0], NULL, 0);
			break;

		case 'g':
			par_c = strtoul (optarg[0], NULL, 0);
			par_h = strtoul (optarg[1], NULL, 0);
			par_s = strtoul (optarg[2], NULL, 0);
			break;

		case 'h':
			par_h = strtoul (optarg[0], NULL, 0);
			break;

		case 'm':
			par_n = strtoul (optarg[0], NULL, 0);
			par_n *= 2048;
			break;

		case 'n':
			par_n = strtoul (optarg[0], NULL, 0);
			break;

		case 'o':
			par_out = optarg[0];
			break;

		case 'q':
			par_quiet = 1;
			break;

		case 's':
			par_s = strtoul (optarg[0], NULL, 0);
			break;

		case 0:
			if (par_out == NULL) {
				par_out = optarg[0];
			}
			else {
				fprintf (stderr, "%s: too many file names (%s)\n",
					argv0, optarg[0]
				);
				return (1);
			}
			break;

		default:
			return (1);
		}
	}

	dsk_adjust_chs (&par_n, &par_c, &par_h, &par_s);

	if (par_quiet == 0) {
		fprintf (stdout, "geometry: %lu/%lu/%lu (%luM at %llu)\n",
			(unsigned long) par_c,
			(unsigned long) par_h,
			(unsigned long) par_s,
			(unsigned long) (par_n / (2 * 1024)),
			(unsigned long long) par_ofs
		);
	}

	if (par_out == NULL) {
		fprintf (stderr, "%s: need a file name\n", argv0);
		return (1);
	}

	if (dsk_create (par_out, par_n, par_c, par_h, par_s, par_ofs)) {
		fprintf (stderr, "%s: create failed (%s)\n", argv0, par_out);
		return (1);
	}

	return (1);
}

static
int main_convert (int argc, char **argv)
{
	int        r;
	char       **optarg;
	const char *par_inp = NULL;
	const char *par_out = NULL;
	const char *par_cow = NULL;
	uint32_t   par_c = 0;
	uint32_t   par_h = 0;
	uint32_t   par_s = 0;
	uint32_t   par_n = 0;
	uint64_t   par_ofs = 0;
	disk_t     *inp, *out, *cow;

	while (1) {
		r = pce_getopt (argc, argv, &optarg, opts_convert);

		if (r == GETOPT_DONE) {
			break;
		}

		if (r < 0) {
			return (1);
		}

		switch (r) {
		case '?':
			print_help();
			return (0);

		case 'V':
			print_version();
			return (0);

		case 'c':
			par_c = strtoul (optarg[0], NULL, 0);
			break;

		case 'C':
			par_min_cluster_size = strtoul (optarg[0], NULL, 0);
			break;

		case 'f':
			par_ofs = strtoull (optarg[0], NULL, 0);
			break;

		case 'g':
			par_c = strtoul (optarg[0], NULL, 0);
			par_h = strtoul (optarg[1], NULL, 0);
			par_s = strtoul (optarg[2], NULL, 0);
			break;

		case 'h':
			par_h = strtoul (optarg[0], NULL, 0);
			break;

		case 'i':
			par_inp = optarg[0];
			break;

		case 'm':
			par_n = strtoul (optarg[0], NULL, 0);
			par_n *= 2048;
			break;

		case 'n':
			par_n = strtoul (optarg[0], NULL, 0);
			break;

		case 'o':
			par_out = optarg[0];
			break;

		case 'q':
			par_quiet = 1;
			break;

		case 's':
			par_s = strtoul (optarg[0], NULL, 0);
			break;

		case 'w':
			par_cow = optarg[0];
			break;

		case 0:
			if (par_inp == NULL) {
				par_inp = optarg[0];
			}
			else if (par_out == NULL) {
				par_out = optarg[0];
			}
			else {
				fprintf (stderr, "%s: too many file names (%s)\n",
					argv0, optarg[0]
				);
				return (1);
			}
			break;

		default:
			return (1);
		}
	}

	if (par_inp == NULL) {
		fprintf (stderr, "%s: need an input file name\n", argv0);
		return (1);
	}

	if (par_out == NULL) {
		fprintf (stderr, "%s: need an output file name\n", argv0);
		return (1);
	}

	inp = dsk_open (par_inp, par_n, par_c, par_h, par_s, par_ofs, 1);
	if (inp == NULL) {
		fprintf (stderr, "%s: can't open input file (%s)\n",
			argv0, par_inp
		);
		return (1);
	}

	if (par_cow != NULL) {
		cow = inp;

		inp = dsk_qed_cow_new (cow, par_cow);

		if (inp == NULL) {
			inp = dsk_cow_new (cow, par_cow);
		}

		if (inp == NULL) {
			dsk_del (cow);
			fprintf (stderr, "%s: can't open COW file (%s)\n",
				argv0, par_cow
			);
			return (1);
		}
	}

	if (par_c == 0) {
		par_c = inp->c;
	}

	if (par_h == 0) {
		par_h = inp->h;
	}

	if (par_s == 0) {
		par_s = inp->s;
	}

	if (par_n == 0) {
		par_n = inp->blocks;
	}

	if (par_quiet == 0) {
		fprintf (stdout, "geometry: %lu/%lu/%lu (%luM at %llu)\n",
			(unsigned long) par_c,
			(unsigned long) par_h,
			(unsigned long) par_s,
			(unsigned long) (par_n / (2 * 1024)),
			(unsigned long long) par_ofs
		);
	}

	if (dsk_create (par_out, par_n, par_c, par_h, par_s, par_ofs)) {
		fprintf (stderr, "%s: can't create output file (%s)\n",
			argv0, par_out
		);
		return (1);
	}

	out = dsk_open (par_out, par_n, par_c, par_h, par_s, par_ofs, 0);
	if (out == NULL) {
		fprintf (stderr, "%s: can't open output file (%s)\n",
			argv0, par_out
		);
		return (1);
	}

	if (dsk_copy (out, inp)) {
		fprintf (stderr, "%s: copy failed\n", argv0);
		return (1);
	}

	dsk_del (out);
	dsk_del (inp);

	return (0);
}

int main (int argc, char **argv)
{
	int  r;
	char **optarg;

	if (argc < 2) {
		print_help();
		return (1);
	}

	argv0 = argv[0];

	while (1) {
		r = pce_getopt (argc, argv, &optarg, opts_main);

		if (r == GETOPT_DONE) {
			break;
		}

		if (r < 0) {
			return (1);
		}

		switch (r) {
		case '?':
			print_help();
			return (0);

		case 'V':
			print_version();
			return (0);

		case 0:
			if (strcmp (optarg[0], "new") == 0) {
				return (main_create (argc, argv));
			}
			else if (strcmp (optarg[0], "create") == 0) {
				return (main_create (argc, argv));
			}
			else if (strcmp (optarg[0], "conv") == 0) {
				return (main_convert (argc, argv));
			}
			else if (strcmp (optarg[0], "convert") == 0) {
				return (main_convert (argc, argv));
			}
			else {
				fprintf (stderr, "%s: unknown command (%s)\n",
					argv0, optarg[0]
				);
				return (1);
			}
			break;

		default:
			return (1);
		}
	}

	return (1);
}
